libImaging/Storage.c

```diff
@@ -339,24 +345,32 @@ ImagingNewBlock(const char *mode, int xsize, int ysize)
 {
     Imaging im;
     Py_ssize_t y, i;
-    Py_ssize_t bytes;
 
     im = ImagingNewPrologue(mode, xsize, ysize);
     if (!im)
         return NULL;
 
-    /* Use a single block */
-    bytes = (Py_ssize_t) im->ysize * im->linesize;
-    if (bytes <= 0)
+    /* We shouldn't overflow, since the threshold defined
+       below says that we're only going to allocate max 4M
+       here before going to the array allocator. Check anyway.
+    */
+    if (im->linesize &&
+        im->ysize > INT_MAX / im->linesize) {
+        /* punt if we're going to overflow */
+        return NULL;
+    }
+
+    if (im->ysize * im->linesize <= 0) {
         /* some platforms return NULL for malloc(0); this fix
            prevents MemoryError on zero-sized images on such
            platforms */
-        bytes = 1;
-    im->block = (char *) malloc(bytes);
+        im->block = (char *) malloc(1);
+    } else {
+        /* malloc check ok, overflow check above */
+        im->block = (char *) calloc(im->ysize, im->linesize);
```

Comment:
Does Pillow have any guarantees about the content of the created images? Maybe we should create a flag which will tell that we are going to use the image internally and fill manually?